components/raftstore-v2/src/operation/command/admin/compact_log.rs

```diff
@@ -0,0 +1,279 @@
+            .compact_entry_cache(std::cmp::min(alive_cache_idx, applied_idx + 1));
+
+        let mut compact_idx = if applied_idx > first_idx
+            && applied_idx - first_idx >= store_ctx.cfg.raft_log_gc_count_limit()
+            || self.approximate_raft_log_size() >= store_ctx.cfg.raft_log_gc_size_limit().0
+        {
+            std::cmp::max(first_idx + (last_idx - first_idx) / 2, replicated_idx)
+        } else if replicated_idx < first_idx
+            || last_idx - first_idx < 3
+            || replicated_idx - first_idx < store_ctx.cfg.raft_log_gc_threshold
+                && self.maybe_skip_compact_log(store_ctx.cfg.raft_log_reserve_max_ticks)
+        {
+            return;
+        } else {
+            replicated_idx
+        };
+        assert!(compact_idx >= first_idx);
+        // Have no idea why subtract 1 here, but original code did this by magic.
+        compact_idx -= 1;
+        if compact_idx < first_idx {
+            return;
+        }
+
+        // Create a compact log request and notify directly.
+        // TODO: move this into a function
+        let term = self.raft_group().raft.raft_log.term(compact_idx).unwrap();
+
+        let mut req = new_admin_request(self.region_id(), self.peer().clone());
+        let mut admin = AdminRequest::default();
+        admin.set_cmd_type(AdminCmdType::CompactLog);
+        admin.mut_compact_log().set_compact_index(compact_idx);
+        admin.mut_compact_log().set_compact_term(term);
+        req.set_admin_request(admin);
+
+        let (ch, _) = CmdResChannel::pair();
+        self.on_admin_command(store_ctx, req, ch);
+
+        self.reset_skip_compact_log_ticks();
+    }
+}
+
+#[derive(Debug)]
+pub struct CompactLogResult {
+    compact_index: u64,
+    compact_term: u64,
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    pub fn propose_compact_log<T>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+        req: RaftCmdRequest,
+    ) -> Result<u64> {
+        let compact_log = req.get_admin_request().get_compact_log();
+        // TODO: add unit tests to cover all the message integrity checks.
+        if compact_log.get_compact_term() == 0 {
+            info!(
+                self.logger,
+                "compact term missing, skip";
+                "command" => ?compact_log
+            );
+            // old format compact log command, safe to ignore.
+            return Err(box_err!(
+                "command format is outdated, please upgrade leader"
+            ));
+        }
+
+        let data = req.write_to_bytes().unwrap();
+        self.propose_with_ctx(store_ctx, data, vec![])
+    }
+}
+
+impl<EK: KvEngine, R: ApplyResReporter> Apply<EK, R> {
+    pub fn apply_compact_log(
+        &mut self,
+        req: &AdminRequest,
+        _log_index: u64,
+    ) -> Result<(AdminResponse, AdminCmdResult)> {
+        Ok((
+            AdminResponse::default(),
+            AdminCmdResult::CompactLog(CompactLogResult {
```

Comment:
Apply state should be changed, and you need log index to correctly store apply state.

Reply:
Which state? apply state in peer is updated in `on_apply_res_compact_log`.

Reply:
`RaftApplyState`. `on_apply_res_compact_log` only updates the memory version, not the disk version.